### PR TITLE
Support local certificates and HTTP proxies

### PR DESCRIPTION
**Pull request by mariovisic**
Hi there 👋 

Found your library as we need to check mTLS certificates using OCSP, thanks for the hard work figuring this all out.

There's a couple of features that I need for our environment:

- We need to specify the proxy host and port for the OCSP checks, but not have them apply globally (so we can't set an environment variable for all net/http calls).
- We already have the client certificate to check as well as the CA bundle, so rather than making a call to a URL to retrieve the chain, instead we can grab the chain from our existing certificates

Here's a patch that implements the changes, I haven't yet included any documentation updates or tests, just keen to firstly hear if you'd be interested in accepting the changes and any feedback.

I also haven't changed the API at all, but it may make sense to add a new class method so that `test` doesn't have two different ways of retrieving the certificate and chain, setting `url` to `nil` feels a bit odd.

Here's the usage when setting the certs and proxy!

```ruby
client_cert = OpenSSL::X509::Certificate.new(File.read("bad_cert.cer"))
ca_certs = OpenSSL::X509::Certificate.load(File.read("ca-bundle-cts.cer"))

SSLTest.test(nil, client_cert: , ca_certs: ca_certs, proxy_host: '...', proxy_port: '...')
```

### CODE REVIEW

**Comment by jarthod** [comment]
Hi @mariovisic :wave: 

Thank you for your contribution, I am interested in accepting those changes after a few updates to the API (and test/readme updates).
1. The proxy change looks ok to me, and you also added it to the CRL method which is good. So this one would just need some tests and documentation in the readme.
2. The API for local cert testing as you said doesn't look great this way and would definitely need two methods I think (something like `#test_cert` and `#test_url` maybe?, `#test` could be an alias for `#test_url` for backward compatibility although this gem doesn't have many users beside me so not a big deal ^^)
It doesn't look like a lot of code would be shared between both methods anyway, a little bit of duplication is fine to me if it makes reading the code simpler, if there's a substantial portion of code to share you can extract it of course.

While we are doing this, and as you seem to understand client and CA certs, I'm not sure how complicated it would be to add the possibility to pass some `ca_certs` to the `#test_url` method in order to validate the server (downloaded) certificate against a private CA, I might use that feature later for https://updown.io ^^ But don't spend your time on this unless you know how to do it and it's easy.

Thanks!

**Comment by mariovisic** [comment]
Neat, i'm excited to get the changes in, I can start working on docs and tests. When I run the current test suite, quite a few tests are failing, i'm assuming because the live data has changed, so I can try send a patch for getting that all green first before merging this PR.

I really like the idea of `test_cert` and `test_url` are you happy for those to be class methods just like `test` is now?

> and as you seem to understand client and CA certs

😬  oh ... I wouldn't say that 😅 


> I'm not sure how complicated it would be to add the possibility to pass some ca_certs to the #test_url method in order to validate the server (downloaded) certificate against a private CA

That would be really straightforward I believe just using the equals operator `==`, from the docs:
https://ruby-doc.org/3.1.3/exts/openssl/OpenSSL/X509/Certificate.html#method-c-load

> Compares the two certificates. Note that this takes into account all fields, not just the issuer name and the serial number.

If you want to compare only some of the fields are the same then we could do that specifically!

If you have some test certificates and a URL then I can verify it works as expected 👍

**Comment by jarthod** [comment]
Thanks for the test update :)
Yes I'm fine with `test_cert` and `test_url` being class methods just like `test `.

I'm not sure it's as simple as comparing certificates, the server certificate will not be "equal" to the private CA cert, it'll be "signed" by the private CA cert. It's another link in the chain, instead of verifying who signed the server certificate from a list of trusted CA installed in the OS, we would check if it was signed by one of the passed private CA cert. But in any case don't worry about this too much and just do what you planned ^^ If I find (or create) a good example I'll see about that later.

**Comment by mariovisic** [comment]
Ooooh, right! I believe you would do the same thing that is in my patch, create a x509 store, then add the CA certificates, then call `store.verify(client_cert)`. It'll error if the `client_ceret` isn't signed by one of the CA certificates added to the store. By default it does not use the CA certificates on the system, you need to call: `#set_default_paths` to use the system ca certs: https://ruby-doc.org/stdlib-3.0.0/libdoc/openssl/rdoc/OpenSSL/X509/Store.html#method-i-set_default_paths

**Comment by mariovisic** [comment]
I've separated the two methods and added the alias, I think the only thing left to do is to add test coverage and documentation and we're good. There are some things that `test_cert` won't do, for example self signed certificates won't be picked up because you're supplying the CA certs, it's just checking against what you provide.

Let me know if anything sticks out that needs changing, otherwise i'll just plug away at the tests, should be good by tomorrow I reckon 👌

File: lib/ssl-test.rb [hunk 1]
```diff
@@ -13,13 +13,14 @@ module SSLTest
   VERSION = -"1.4.1"
 
   class << self
-    def test url, open_timeout: 5, read_timeout: 5, redirection_limit: 5
+    def test_url url, open_timeout: 5, read_timeout: 5, proxy_host:nil, proxy_port: nil, redirection_limit: 5, ca_certs: []
```
Comment (jarthod) [inline lib/ssl-test.rb]: The `ca_certs` parameter is not used here, it's here for the eventual other feature?

File: lib/ssl-test/ocsp.rb [hunk 2]
```diff
@@ -82,7 +82,7 @@ def follow_ocsp_redirects(uri, data, open_timeout: 5, read_timeout: 5, redirecti
         @logger&.debug { "SSLTest   + OCSP: 200 OK (#{http_response.body.bytesize} bytes)" }
         [http_response.body, nil]
       when Net::HTTPRedirection
-        follow_ocsp_redirects(URI(http_response["location"]), data, open_timeout: open_timeout, read_timeout: read_timeout, redirection_limit: redirection_limit - 1)
+        follow_ocsp_redirects(URI(http_response["location"]), data, open_timeout: open_timeout, read_timeout: read_timeout, rproxy_host: proxy_host, proxy_port: proxy_port, edirection_limit: redirection_limit - 1)
```
Comment (jarthod) [inline lib/ssl-test/ocsp.rb]: Some typos here (rproxy_host, ederection_limit): if you can, make sure the specs would catch those ^^

**Comment by jarthod** [comment]
Thanks! I just left 2 small comments. Yes for self-signed certificates it makes sense, all custom CA certs can be qualified as "self-signed" as long as you don't have the parent CA ^^ (and if you have it they are valid).

File: lib/ssl-test.rb (hunk 1 above)
Comment (mariovisic) [reply lib/ssl-test.rb]: This was for the feature that you were describing above, but I can do that in a followup PR 👍

File: lib/ssl-test/ocsp.rb (hunk 2 above)
Comment (mariovisic) [reply lib/ssl-test/ocsp.rb]: Neat, have fixed and added proxy tests which cover this 👍

**Comment by mariovisic** [comment]
Thanks for being patient, this is ready for review now, I've updated all the tests and the documentation on the readme!

File: spec/ssl-test_spec.rb [hunk 3]
```diff
@@ -8,7 +10,16 @@
...
 
-  describe '.test' do
+  let(:proxy_thread) { nil }
+
+
+  after(:each) do
+    if proxy_thread
+      proxy_thread.kill
+    end
+  end
```
Comment (jarthod) [inline spec/ssl-test_spec.rb]: I would probably shorten this as:
```suggestion
  after(:each) { proxy_thread&.kill }
```

File: spec/ssl-test_spec.rb [hunk 4]
```diff
@@ -235,4 +277,189 @@
...
+
+      context 'when the proxy is active' do
+        it 'uses the provided http proxy' do
+          proxy_thread
+          sleep 0.1 # wait for the proxy to start!
+
+          cert = OpenSSL::X509::Certificate.new(File.read(File.join(__dir__, 'fixtures/google_com_client.pem')))
+          ca_bundle = OpenSSL::X509::Certificate.load(File.read(File.join(__dir__, 'fixtures/google_com_ca_bundle.pem')))
+
+          valid, error, cert = SSLTest.test_cert(cert, ca_bundle, proxy_host: '127.0.0.1', proxy_port: 8080)
```
Comment (jarthod) [inline spec/ssl-test_spec.rb:451]: I'm not sure how to change it, but this is actually not testing that we use the proxy, if we remove all the code using "proxy_host" and "proxy_port" and just leave the unused parameters, this test would still pass.
Can you check if the WEBrick::HTTPProxyServer has something which changes on request (a counter or something) that we could use to make sure ? Otherwise I guess you can use the logger or access-log (in memory like with a StringIO or something) to verify the expected request were all proxied ?
(some thing above in the `test_url` specs of course)

File: lib/ssl-test.rb [hunk 5]
```diff
@@ -81,6 +105,28 @@ def logger= logger
 
     private
 
+    def parsed_message(revoked, revocation_date, message)
```
Comment (jarthod) [inline lib/ssl-test.rb]: Just a note about naming here, this function name is not explicit enough IMO, maybe something like `revocation_message` ? (we're not really parasing anything here, just generating the revocation error message if any)

**Review by jarthod — COMMENTED** [review]
Pretty good thanks ! I left a few comments.
I did not expect this many tests and fixtures (I see you copied all the cases from the URL check), not sure if they are all relevant to a client cert test but that's fine for me.

File: spec/ssl-test_spec.rb (hunk 3 above)
Comment (mariovisic) [reply spec/ssl-test_spec.rb]: Great, thank you :)

File: spec/ssl-test_spec.rb (hunk 4 above)
Comment (mariovisic) [reply spec/ssl-test_spec.rb:451]: That is true, I think combined with the following test not being able to connect, it does indirectly show the proxy is working, but you're right, if I can spy on the requests to the proxy (in theory this should work fine, but unsure if the thread will get in the way), then we can verify a request has gone to the proxy 👍

File: lib/ssl-test.rb (hunk 5 above)
Comment (mariovisic) [reply lib/ssl-test.rb]: Good suggestion, thank you, updated in 2ac80b42188c33f995ca2245954c9d8760fbbf07

**Comment by mariovisic** [comment]
> Pretty good thanks ! I left a few comments. I did not expect this many tests and fixtures (I see you copied all the cases from the URL check), not sure if they are all relevant to a client cert test but that's fine for me.

Ah right, I thought I'd only copied the relevant ones, I'm happy to remove tests that aren't testing anything useful

**Comment by mariovisic** [comment]
I've updated the tests to check that the proxy received requests 👌  and it's all working well, checkout commit: 46dd72d389ad27b2c3f0ac6919463cf50622f366

I think this is ready for review again :)

File: spec/ssl-test_spec.rb (hunk 4 above)
Comment (mariovisic) [reply spec/ssl-test_spec.rb:451]: Updated in 46dd72d389ad27b2c3f0ac6919463cf50622f366

File: spec/ssl-test_spec.rb [hunk 6]
```diff
@@ -176,6 +183,41 @@
...
+          $proxy.start
+        end
+      end
+
+      context 'when the proxy is active' do
+        it 'uses the provided http proxy' do
+          proxy_thread
+          sleep 0.1 # wait for the proxy to start!
+
+          allow($proxy).to receive(:do_GET).and_call_original
```
Comment (jarthod) [inline spec/ssl-test_spec.rb]: Thanks for the assertion, quick note: I believe you can do it in a single line with:
```suggestion
          expect($proxy).to receive(:do_GET).twice.and_call_original
```
(no need for the `have_received` in this case) but to be verified ^^

**Review by jarthod — APPROVED** [review]
Thanks, looks good to me 👍 
I let you have a look at the final comment and I'll merge after that.

File: spec/ssl-test_spec.rb (hunk 6 above)
Comment (mariovisic) [reply spec/ssl-test_spec.rb]: Yeah totally, that will work great, I usually avoid using the mock expectation that `expect().to receive` creates  because the test structure is harder to follow ... eg: with the existing code it's `setup -> execution -> expectation`, which is the same pattern for tests that aren't isolating with stubs/mocks/spies etc...  where as using the mock the steps are: `setup -> expectation -> execution` which is harder to follow as it's different to other tests:


Here's a test without any external dependencies
```ruby
# setup (this could be in a let as well)
user = User.new

# execution
result = user.name

# expectation
expect(result).to eq('Adrien')
```

Here's an example using a spy via `allow`
```ruby
# setup
user = User.new
allow(MyExternalDependency).to receive(:test)

# execution
user.name

# expectation
expect(MyExternalDependency) to have_received(:test)
```

Here's the same with the mock example via `expect ... to receive`
```ruby
# setup
user = User.new

# expectation
expect(MyExternalDependency).to receive(:test)

# execution
user.name
```

Comment (mariovisic) [reply spec/ssl-test_spec.rb]: I don't have a strong opinion on one or the other, just wanted to go through my rationale, happy to change it 👍

Comment (jarthod) [reply spec/ssl-test_spec.rb]: Indeed, I am used to the having the expectation first with rspec mocks so that does not bother me much but I understand the rational and I agree it's easier to read/understand the way you wrote it. Let's keep it that way!

In that same spirit though, wouldn't it be a good idea to move the `allow($proxy).to receive(:do_GET).and_call_original` line inside the proxy setup `let` ? (as it's not serving any purpose for understanding of the test, it's just required to setup rspec mock so we can count the number of call at the end)

Comment (mariovisic) [reply spec/ssl-test_spec.rb]: Good idea, added in 137994a3225663d627d077eb9f50dd20fc0d5005  I moved the `sleep` too as it's part of setting up the proxy!

Comment (jarthod) [reply spec/ssl-test_spec.rb]: Nice! and sorry but I just noticed the "let(:proxy_thread)" block is present twice and identical? can't we move it up to the top block?

Comment (mariovisic) [reply spec/ssl-test_spec.rb]: Yes, though will need to solve the problem of killing it once it's done. We have this code: `proxy_thread&.kill` to stop the proxy which will boot the proxy when calling `proxy_thread` if we moved `proxy_thread` to the top.

We could boot the proxy once for all of the tests, though we'd need to reset the stub each time.

killing the proxy could be moved to the test body but it won't run if the test runs into an exception.

Comment (mariovisic) [reply spec/ssl-test_spec.rb]: ... it'll boot the proxy as the `let` is lazily loaded

Comment (jarthod) [reply spec/ssl-test_spec.rb]: Ah yep ok, let's keep it that way then :)

File: spec/ssl-test_spec.rb [hunk 7]
```diff
@@ -176,6 +183,43 @@
...
+      end
+
+      context 'when the proxy is not reachable' do
+        it 'returns a http error' do
+          valid, error, cert = SSLTest.test("https://updown.io", proxy_host: '127.0.0.1', proxy_port: 55000)
+          expect(error).to include('(Connection refused - connect(2) for "127.0.0.1" port 55000)')
+          expect(valid).to be_nil
+          expect(cert).to be_nil
+        end
+      end
```
Comment (jarthod) [inline spec/ssl-test_spec.rb:221]: But if I'm correct, it means this test will actually also start the proxy during the teardown (due to `after(:each) { proxy_thread&.kill }`) while not needing it at all during the test? Not a big deal but maybe the proxy should be in the `context 'when the proxy is active' do` then?

File: spec/ssl-test_spec.rb [hunk 8]
```diff
@@ -235,4 +279,195 @@
...
+        it 'returns a http error' do
+          cert = OpenSSL::X509::Certificate.new(File.read(File.join(__dir__, 'fixtures/google_com_client.pem')))
+          ca_bundle = OpenSSL::X509::Certificate.load(File.read(File.join(__dir__, 'fixtures/google_com_ca_bundle.pem')))
+
+          valid, error, cert = SSLTest.test_cert(cert, ca_bundle, proxy_host: '127.0.0.1', proxy_port: 55000)
+          expect(error).to include('(Connection refused - connect(2) for "127.0.0.1" port 55000)')
+          expect(valid).to be_nil
+          expect(cert).to eq(cert)
+        end
+      end
```
Comment (jarthod) [inline spec/ssl-test_spec.rb:470]: Same thing here of course

**Comment by mariovisic** [comment]
good catch! Totally. You’re right, we should move the let to be inside the specific context. I’m away from my PC so can do this with tonight, I’m also happy if you want to commit the change if that’s easier!On 22 Nov 2025, at 10:20 pm, Adrien Rey-Jarthon ***@***.***> wrote:﻿
@jarthod commented on this pull request.



In spec/ssl-test_spec.rb:
> +
+          expect($proxy).to have_received(:do_GET).once
+        end
+      end
+
+      context 'when the proxy is not reachable' do
+        it 'returns a http error' do
+          cert = OpenSSL::X509::Certificate.new(File.read(File.join(__dir__, 'fixtures/google_com_client.pem')))
+          ca_bundle = OpenSSL::X509::Certificate.load(File.read(File.join(__dir__, 'fixtures/google_com_ca_bundle.pem')))
+
+          valid, error, cert = SSLTest.test_cert(cert, ca_bundle, proxy_host: '127.0.0.1', proxy_port: 55000)
+          expect(error).to include('(Connection refused - connect(2) for "127.0.0.1" port 55000)')
+          expect(valid).to be_nil
+          expect(cert).to eq(cert)
+        end
+      end

Same thing here of course

**Comment by jarthod** [comment]
I just merged and changed the spec in the merge commit :)
I also released `1.5.0` with this change.

**Comment by mariovisic** [comment]
> I just merged and changed the spec in the merge commit :) I also released `1.5.0` with this change.

Thank you :) what great news, thanks for collaborating to get this over the line!
